src/frequenz/sdk/data_ingestion/resampling/ring_buffer.py

```diff
@@ -0,0 +1,63 @@
+"""
+Ring buffer for storing ordered streamed component data for resampling purposes.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+import logging
+from collections import deque
+from typing import Deque, Generic, List, TypeVar
+
+logger = logging.Logger(__name__)
+
+MAX_BUFFER_SIZE = 1000
+
+T = TypeVar("T")
+
+
+class RingBuffer(Generic[T]):
+    """Ring buffer for storing items."""
+
+    def __init__(
+        self,
+        size: int,
+    ) -> None:
+        """Initialize RingBuffer.
+
+        Args:
+            size: max number of items to be stored in the buffer
+
+        Raises:
+            ValueError: if provided `buffer_size` is smaller than 0 or greater than
+                `MAX_BUFFER_SIZE`
+        """
+        if size < 0 or size > MAX_BUFFER_SIZE:
+            raise ValueError(
+                f"Buffer size must be an integer in the (0, {MAX_BUFFER_SIZE}) range."
+            )
+
+        self._size = size
+        self._buffer: Deque[T] = deque()
+
+    @property
+    def items(self) -> List[T]:
+        """Get all the items in the ring buffer.
+
+        Returns:
+            All the items in the ring buffer.
+        """
+        return list(self._buffer)
+
+    def insert(self, item: T) -> None:
+        """Insert a new item into the ring buffer.
+
+        Args:
+            item: item to be added to the buffer
+        """
+        if len(self._buffer) >= self._size:
+            self._buffer.popleft()
```

Comment:
From doc: https://docs.python.org/3/library/collections.html#collections.deque

deque has max_len argument, so why you do it manually?

Reply:
Because if I change the `self._size` in the `def resize(self, new_size: int)` method, it will be properly taken into account here, without touching the underlying collection.

If I use `deque(maxlen=size)` in the constructor, I would have to recreate the underlying collection with a new size and copy over the elements. But actually now that I think of it, it might be a better solution to recreate it, because someone can downsize the ring buffer.

Reply:
>Since deque provides more flexibility we probably just have to encourage people to use a deque where a ring buffer is needed.

Agreed. Doesn't look like this RingBuffer module achieves anything significant other than just being a wrapper over `Deque`.